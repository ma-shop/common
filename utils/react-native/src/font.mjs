import { round, ui } from '@ma-shop/utils'

/// @name em
/// @description This function is used to mimic the `em` functionality of css
/// em's allow us to code things responsively in a very easy way
/// @arg {number} value - The em value you want to get
/// @returns {unit} - the px value based on the baseUnit size
export function em (value = 1, base) {
  return round(value * (base || ui.fontSize))
}

/// @name px
/// @description this will convert your pixel value to em's
/// @arg {number} value - the pixel to compute
/// @returns {unit} - the em value based on the base unit
export function px (value, base) {
  return round(value / (base || ui.fontSize))
}
