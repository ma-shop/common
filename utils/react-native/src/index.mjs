// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
// !!! This file is generated do not modify it manually because it will be overwritten
// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

export * from './get-device-locale'
export * from './interactions'
export * from './is'
export * from './navigation'
export * from './shapes'
