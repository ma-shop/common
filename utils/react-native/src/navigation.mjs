export const nav = (() => {
  class Navigation {
    resetRoute = null

    set (obj) {
      this.navigation = obj
    }

    ///# @name to
    ///# @description navigate to a page
    ///# https://reactnavigation.org/docs/en/navigation-prop.html#navigate-link-to-other-screens
    ///# @example
    ///# nav.to('ProductDetail', { prodId: 13009 })
    to (...args) {
      this.navigation.navigate(...args)
    }

    ///# @name back
    ///# @description navigate back to a page
    ///# https://reactnavigation.org/docs/en/navigation-prop.html#goback-close-the-active-screen-and-move-back
    ///# @arg {null, string} screen [null] - The screen to go back to.
    ///# If it's null it will just go back to the previous screen
    ///# @alias `nav.close()`
    ///# @example
    ///# // goes back to previous page
    ///# nav.back()
    ///# // goes back until it gets to the product detail page
    ///# nav.back('ProductDetail')
    back = (screen = null) => this.navigation.goBack(screen)

    close = () => this.back(null)

    ///# @name replace
    ///# @description replace the current page without adding it to the stack
    ///# it uses the same arguments as `this.nav.to`
    ///# https://reactnavigation.org/docs/en/navigation-prop.html#replace
    ///# @example
    ///# this.nav.replace('ProductDetail', { prodId: 13009 })
    replace (...args) {
      this.navigation.replace(...args)
    }

    ///# @name setParams
    ///# @description sets the params on the current view
    ///# https://reactnavigation.org/docs/en/navigation-prop.html#setparams-make-changes-to-route-params
    ///# @example
    ///# this.nav.setParams({})
    setParams (params) {
      this.navigation.setParams(params)
    }

    ///# @name setOptions
    ///# @description sets the options on the current view
    ///# https://reactnavigation.org/docs/navigation-prop/#setoptions
    ///# @example
    ///# this.nav.setOptions({})
    setOptions (options) {
      this.navigation.setOptions(options)
    }

    ///# @name isFocused
    ///# @description lets you know if the current view is focused or not
    ///# https://reactnavigation.org/docs/en/navigation-prop.html#isfocused-query-the-focused-state-of-the-screen
    ///# @example
    ///# this.nav.isFocused
    ///# @getter
    get isFocused () {
      return this.navigation.isFocused()
    }

    ///# @name on
    ///# @description lets you know if the current view is focused or not
    ///# @arg {stirng} event - one of these events `willBlur`, `willFocus`, `didFocus`, `didBlur`
    ///# https://reactnavigation.org/docs/en/navigation-prop.html#addlistener-subscribe-to-updates-to-navigation-lifecycle
    ///# @example
    ///# useEffect(() => {
    ///#   const unsubscribe = this.nav.on('willBlur', (e) => {
    ///#     console.log('e:', e)
    ///#   })
    ///#
    ///#   return unsubscribe
    ///# }, [nav.navigation])
    ///# @returns {fn} The unsubscribe function to remove the event listener
    on (event, cb) {
      // willBlur, willFocus, didFocus, didBlur
      return this.navigation.addListener(event, cb)
    }

    ///# @name dispatch
    ///# @describe dispatches navigation actions
    ///# @args {object} ...args - same as navigation dispatch
    dispatch (...args) {
      return this.navigation.dispatch(...args)
    }

    ///# @name reset
    ///# @description resets the navigation
    ///# @arg {string} name - The route you want to reset it to
    ///# @note you can set a default route using `nav.setResetRoute`
    reset (name) {
      return this.navigation.reset({
        index: 0,
        routes: [ { name: name || this.resetRoute } ],
      })
    }

    ///# @name setResetRoute
    ///# @description Sets the default reset route
    ///# @arg {string} name - the route to reset it to
    setResetRoute (name) {
      this.resetRoute = name
    }
  }

  return new Navigation()
})()
