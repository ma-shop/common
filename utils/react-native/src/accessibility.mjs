import { PixelRatio } from 'react-native'
import { clamp } from '@ma-shop/utils'

/// @name accessibilityStates
/// @author Tyler Benton
/// @description This is used to handle the accessibilityStates prop in a better way
/// since dealing with an array of strings is a weird implementation
/// @arg {object} obj = The states that you want to set
/// @example
///
/// render () {
///   const { isDisabled } = this.state
///
///   return (
///     <View
///       accessible
///       accessibilityStates={accessibilityStates({ isDisabled })}
///     />
///   )
/// }
export function accessibilityStates (obj) {
  return Object.keys(obj)
    .map((key) => obj[key] && key.replace(/^is/, '').toLowerCase())
    .filter(Boolean)
}


/// @name fontScale
/// @author Tyler Benton
/// @description This allows you to fake the font scaling if it's needed.
/// This is typically used with Images but it can be used in other cases if needed.
/// @note When you change the font scale in the accessibility inspector it will not show changes
/// on the screen like the actual fonts show. To see changes take affect you will have to
/// reload the screen.
export function fontScale ({
  allowFontScaling = true,
  noFontScaling = false,
  maxFontSizeMultiplier = 2.25,
  minimumFontScale = 1,
} = {}) {
  if (!allowFontScaling || noFontScaling) return 1
  return clamp(PixelRatio.getFontScale(), minimumFontScale, maxFontSizeMultiplier)
}
