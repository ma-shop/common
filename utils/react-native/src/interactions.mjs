import { InteractionManager } from 'react-native'
import { delay } from '@ma-shop/utils'

/// @name interactions
/// @author Tyler Benton
/// @description This function is used to wait for other interactions to finish
/// before you run a specific interaction. It allows animtations to run smooth
/// because there's not as many animations running at a time
/// @arg {number} max [1000] - The max time you wait for interactions to finish
export function interactions (max = 1000) {
  return Promise.race([
    new Promise(InteractionManager.runAfterInteractions),
    delay(max),
  ])
}
