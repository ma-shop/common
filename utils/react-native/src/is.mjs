import { Platform } from 'react-native'
import { is } from '@ma-shop/is'


/// @name ios
/// @description This will check if the current device is on ios
/// @returns {boolean}
is.ios = () => Platform.OS === 'ios'

/// @name android
/// @description This will check if the current device is on android
/// @arg {array, string} - The types to test against
/// @returns {boolean}
is.android = () => Platform.OS === 'android'
