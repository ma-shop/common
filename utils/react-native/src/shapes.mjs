import { Text, ViewPropTypes } from 'react-native'
import { shapes } from '@ma-shop/react'
import PropTypes from 'prop-types'


shapes.textStyle = PropTypes.oneOfType([
  Text.propTypes.style,
  PropTypes.arrayOf(Text.propTypes.style),
])

shapes.viewStyle = PropTypes.oneOfType([
  ViewPropTypes.style,
  PropTypes.arrayOf(ViewPropTypes.style),
])
