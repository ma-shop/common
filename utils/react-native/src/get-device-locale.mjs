import { NativeModules } from 'react-native'
import { is } from '@ma-shop/is'

/// @name getDeviceLocale
/// @description This function will get the current locale for where the device is.
/// @returns {string} - eg. en-US
// @todo Move this to the @ma-shop/react-native library
export function getDeviceLocale () {
  let deviceLocale

  if (is.ios()) {
    deviceLocale = NativeModules.SettingsManager.settings.AppleLocale
    if (!deviceLocale) {
      // iOS 13 workaround, take first of AppleLanguages array
      [ deviceLocale ] = NativeModules.SettingsManager.settings.AppleLanguages
    }
  } else {
    deviceLocale = NativeModules.I18nManager.localeIdentifier
  }

  // device locale isn't available; return en-us
  if (!deviceLocale) return 'en-US'

  const localePieces = deviceLocale.split('_')

  if (localePieces.length !== 2) return 'en-US'

  const language = localePieces[0].toLowerCase()
  const country = localePieces[1].toUpperCase()
  return `${language}-${country}`
}
