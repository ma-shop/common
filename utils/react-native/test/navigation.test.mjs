import { nav } from '../src/navigation'


const mockNavigation = (overrides = {}) => {
  const navigation = {
    navigate: jest.fn(),
    goBack: jest.fn(),
    replace: jest.fn(),
    setParams: jest.fn(),
    setOptions: jest.fn(),
    isFocused: jest.fn(() => true),
    addListener: jest.fn((event, callback) => {
      return () => navigation.removeListener(event, callback)
    }),
    removeListener: jest.fn(),
    dispatch: jest.fn(),
    reset: jest.fn(),
    ...overrides,
  }

  nav.set(navigation)
}

describe('navigation', () => {
  beforeEach(() => {
    mockNavigation()

    // reset this before every test
    nav.resetRoute = null
  })

  test('set', () => {
    nav.set('woohoo')

    expect(nav.navigation).toBe('woohoo')
  })

  test('to', () => {
    nav.to('Login', { id: 'woohoo' })

    expect(nav.navigation.navigate).toBeCalledWith('Login', { id: 'woohoo' })
  })

  test('back', () => {
    nav.back()

    expect(nav.navigation.goBack).toBeCalledWith(null)

    nav.back('Login')

    expect(nav.navigation.goBack).toBeCalledWith('Login')
  })

  test('close', () => {
    nav.close()

    expect(nav.navigation.goBack).toBeCalledWith(null)

    nav.close('Login')

    expect(nav.navigation.goBack).toBeCalledWith(null)
  })

  test('replace', () => {
    nav.replace('Login', { id: 'woohoo' })

    expect(nav.navigation.replace).toBeCalledWith('Login', { id: 'woohoo' })
  })

  test('setParams', () => {
    nav.setParams({ id: 'woohoo' })

    expect(nav.navigation.setParams).toBeCalledWith({ id: 'woohoo' })
  })

  test('setOptions', () => {
    nav.setOptions({ id: 'woohoo' })

    expect(nav.navigation.setOptions).toBeCalledWith({ id: 'woohoo' })
  })

  test('isFocused', () => {
    expect(nav.isFocused).toBeTruthy()
    expect(nav.navigation.isFocused).toBeCalled()
  })

  test('on', () => {
    const fn = () => 'woohoo'
    const unsubscribe = nav.on('focus', fn)

    expect(nav.navigation.addListener).toBeCalled()
    expect(unsubscribe).toEqual(expect.any(Function))

    unsubscribe()
    expect(nav.navigation.removeListener).toBeCalled()
  })

  test('dispatch', () => {
    nav.dispatch('woohoo')

    expect(nav.navigation.dispatch).toBeCalledWith('woohoo')
  })

  test('setResetRoute', () => {
    expect(nav.resetRoute).toBe(null)

    nav.setResetRoute('DefaultRoute')

    expect(nav.resetRoute).toBe('DefaultRoute')
  })

  describe('reset', () => {
    beforeEach(() => {
      nav.setResetRoute('DefaultRoute')
    })

    test('with default', () => {
      nav.reset()

      expect(nav.navigation.reset.mock.calls[0]).toMatchSnapshot()
    })

    test('with override', () => {
      nav.reset('NoneDefaultRoute')

      expect(nav.navigation.reset.mock.calls[0]).toMatchSnapshot()
    })
  })
})
