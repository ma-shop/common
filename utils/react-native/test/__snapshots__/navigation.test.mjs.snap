// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigation reset with default 1`] = `
Array [
  Object {
    "index": 0,
    "routes": Array [
      Object {
        "name": "DefaultRoute",
      },
    ],
  },
]
`;

exports[`navigation reset with override 1`] = `
Array [
  Object {
    "index": 0,
    "routes": Array [
      Object {
        "name": "NoneDefaultRoute",
      },
    ],
  },
]
`;
